crates/codegen/src/yul/isel/inst_order.rs

```diff
@@ -333,6 +388,168 @@ impl<'a> InstSerializer<'a> {
+            scrutinee,
+            table: analyzed_table,
+            default: analyzed_default,
+            merge_block: self.select_switch_merge_block(
+                &merge_block_cands,
+                table.iter().map(|(_, d)| d).chain(default),
+            ),
+        }
+    }
+
+    fn find_merge_block_candidates(
+        &self,
+        branch_inst_bb: BasicBlockId,
+        branch_dest_bb: BasicBlockId,
+    ) -> Vec<BasicBlockId> {
+        if self.domtree.dominates(branch_dest_bb, branch_inst_bb) {
+            return vec![];
+        }
+
+        // a block `cand` can be a candidate of a `merge` block iff
+        // 1. `cand` is a dominance frontier of `branch_dest_bb`.
+        // 2. `cand` is NOT a dominator of `branch_dest_bb`.
+        // 3. `cand` is NOT a "merge" block of parent `if` or `switch`.
+        // 4. `cand` is NOT a "loop_exit" block of parent `loop`.
+        match self.df.frontiers(branch_dest_bb) {
+            Some(cands) => cands
+                .filter(|cand| {
+                    !self.domtree.dominates(*cand, branch_dest_bb)
+                        && Some(*cand)
+                            != self
+                                .scope
+                                .as_ref()
+                                .and_then(Scope::branch_merge_block_recursive)
+                        && Some(*cand) != self.scope.as_ref().and_then(Scope::loop_exit_recursive)
+                })
+                .collect(),
+            None => vec![],
+        }
+    }
+
+    /// Each destination block of `switch` instruction could have multiple
+    /// candidates for the merge block because arm bodies can have multiple
+    /// predecessors, e.g., `default` arm.
+    /// So we need a heuristic to select the merge block from candidates.
+    ///
+    /// First, if one of the dominance frontiers of switch dests is a parent
+    /// merge block, then we stop searching the merge block because the parent
+    /// merge block should be the subsequent codes after the switch in terms of
+    /// high-level flow structure like Fe or yul.
+    ///
+    /// If no parent merge block is found, we start scoring the candidates by
+    /// the following function.
+    ///
+    /// The scoring function `F` is defined as follows:
+    /// 1. The initial score of each candidate('cand_bb`) is number of
+    /// predecessors of the candidate.
+    ///
+    /// 2. Find the `top_cand` of each `cand_bb`. `top_cand` can be found by
+    /// [`Self::try_find_top_cand`] method, see the method for details.
+    ///
+    /// 3. If `top_cand` is found, then add the `cand_bb` score to the
+    /// `top_cand` score, then set 0 to the `cand_bb` score.
+    ///
+    /// After the scoring, the candidates with the highest score will be
+    /// selected.
+    fn select_switch_merge_block(
+        &self,
+        cands: &IndexSet<BasicBlockId>,
+        dests: impl Iterator<Item = BasicBlockId>,
+    ) -> Option<BasicBlockId> {
```

Comment:
This heuristic is necessary to select a `switch`'s exit or merge block because multiple merge block candidates cannot be distinguished only by looking at a `MIR` CFG. This is because of the existence of a `default` arm.
Please consider the following example and the corresponding MIR. 

```rust
pub enum MyEnum {
    Unit
    UnitTuple()
    Tuple(u32, u256)
}

pub enum Tuple2{
    Value(MyEnum, MyEnum)
}


contract Foo {
    pub fn bar() -> u256 {
        let tup: Tuple2 = Tuple2::Value(MyEnum::Tuple(1, 2), MyEnum::UnitTuple())
        let res: u256 = 0
        match tup {
            Tuple2::Value(MyEnum::Unit, MyEnum::Tuple(x, y)) | Tuple2::Value(MyEnum::Tuple(x, y), MyEnum::UnitTuple()) => {
                res = u256(x) + y
            }
            _ => {
                return 0
            }
        }
        
        return res
    }
}
```
![graphviz](https://user-images.githubusercontent.com/6376004/188288017-b77e9691-0f4c-4270-882e-3774a9c05a76.png)

Here, the "original" exit block is `bb1`(or `bb5` which can be seen as the exit block because it's the immediate dominator of `bb1`). But only by looking at the MIR we can't distinguish it from `bb7`, because they are both in the dominance frontier of `bb3` and `bb7`.

NOTE:
1. This algorithm is deterministic.
2. This algorithm doesn't change the semantics of the original code even if it fails to find the original exit block. But if it fails to find the original exit block, the generated yul would be much greater than the optimal code. 
3. If there is no `return` in all the arms, this algorithm can certainly find the exit block.